services/CurrentTime/source/CurrentTimeService.cpp

```diff
@@ -0,0 +1,227 @@
+    _current_time_handler = handler;
+}
+
+time_t CurrentTimeService::get_time() const
+{
+    time_t epoch_time = time(nullptr);
+
+    return epoch_time + _time_offset;
+}
+
+void CurrentTimeService::set_time(time_t host_time, uint8_t adjust_reason)
+{
+    time_t epoch_time = time(nullptr);
+
+    _time_offset = host_time - epoch_time;
+
+    update_current_time_value(adjust_reason);
+}
+
+void CurrentTimeService::update_current_time_value(const uint8_t adjust_reason) {
+    time_t local_time = get_time();
+
+    struct tm *local_time_tm = localtime(&local_time);
+
+    CurrentTime current_time(local_time_tm);
+
+    current_time.adjust_reason = adjust_reason;
+
+    _ble.gattServer().write(_current_time_char.getValueHandle(),
+                            reinterpret_cast<const uint8_t *>(&current_time),
+                            CURRENT_TIME_CHAR_VALUE_SIZE);
+
+    if (_event_queue_handle != 0) {
+        _event_queue.cancel(_event_queue_handle);
+        _event_queue_handle  = 0;
+    }
+
+    start_periodic_time_update();
+}
+
+void CurrentTimeService::start_periodic_time_update() {
+    if (_event_queue_handle == 0) {
+        _event_queue_handle = _event_queue.call_in(UPDATE_TIME_PERIOD, [this] {
+            _event_queue_handle = 0;
+            update_current_time_value(EXTERNAL_REFERENCE_TIME_UPDATE);
+        });
+    }
+}
+
+void CurrentTimeService::onCurrentTimeRead(GattReadAuthCallbackParams *read_request)
+{
+    time_t local_time = get_time();
+    CurrentTime local_current_time(localtime(&local_time));
+
+    if (local_current_time.valid()) {
+        _current_time = local_current_time;
+        read_request->data = reinterpret_cast<uint8_t *>(&_current_time);
+        read_request->len  = CURRENT_TIME_CHAR_VALUE_SIZE;
+        read_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+    } else {
+        read_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_UNLIKELY_ERROR;
+    }
+}
+
+void CurrentTimeService::onCurrentTimeWritten(GattWriteAuthCallbackParams *write_request)
+{
+    CurrentTime input_time(write_request->data);
+
+    if (!input_time.valid()) {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_OUT_OF_RANGE;
+        return;
+    }
+
+    struct tm remote_time_tm{};
+    input_time.to_tm(&remote_time_tm);
+    time_t remote_time = mktime(&remote_time_tm);
+
+    set_time(remote_time, input_time.adjust_reason);
+
+    if (_current_time_handler) {
+        _current_time_handler->on_current_time_changed(remote_time, input_time.adjust_reason);
+    }
+
+    if (input_time.fractions256) {
+        write_request->authorizationReply = (GattAuthCallbackReply_t) DATA_FIELD_IGNORED;
+    } else {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+    }
+}
+
+CurrentTimeService::CurrentTime::CurrentTime(const uint8_t *data)
+{
+    // FIXME: 2bit transmitted in little endian
+    year          = *data | (*(data + 1) << 8);
```

Comment:
the data is already in the correct order

Reply:
this is not resolved, this whole thing could just be a memcpy

Reply:
I think it is important to keep it in that form so ww know what to change if we add BE support. 
Not a blocker for me at the moment.

Reply:
OK, but when we do have BE support then it will be a memcpy - which is my point but yeah, not a blocker

Reply:
Yes, we can optimize it when that support is added. That is internal details today and a small micro optimization. It is fine to keep it like it today.